src/contrib/mlir/backend/cpu/cpu_backend.cpp

```diff
@@ -0,0 +1,274 @@
+
+static llvm::cl::opt<bool>
+    clEnableAffineLoopTiling("ngraph-affine-loop-tile",
+                             llvm::cl::init(false),
+                             llvm::cl::desc("Enable loop tiling optimization in Affine dialect"));
+
+static llvm::cl::opt<unsigned>
+    clLoopTilingCacheLevel("ngraph-affine-loop-tile-cache-level",
+                           llvm::cl::init(2),
+                           llvm::cl::desc("Cache level to which to apply affine loop tiling."));
+
+static llvm::cl::opt<unsigned> clLoopTilingCacheSize(
+    "ngraph-affine-loop-tile-cache-size",
+    llvm::cl::init(0),
+    llvm::cl::desc(
+        "Cache size to use in affine loop tiling. If not zero, it overrides the cache-size "
+        "inferred from the host CPU using for the cache level specified by "
+        "-ngraph-loop-tile-cache-level."));
+
+using namespace ngraph::runtime::ngmlir;
+
+// Default optimization level.
+llvm::CodeGenOpt::Level MLIRCPUBackend::mlirOptLevel = llvm::CodeGenOpt::Level::Aggressive;
+
+std::unique_ptr<llvm::TargetMachine> MLIRCPUBackend::targetMachine;
+
+bool MLIRCPUBackend::initialized = false;
+
+/// Creates target machine for current host.
+static llvm::Expected<std::unique_ptr<llvm::TargetMachine>>
+    createDefaultTargetMachine(unsigned optLevel)
+{
+    auto machineBuilder = llvm::orc::JITTargetMachineBuilder::detectHost();
+    if (!machineBuilder)
+    {
+        return machineBuilder.takeError();
+    }
+
+    // Relocation model and code model are kept to default values. CodeGen optimization level
+    // matches LLVM recommendations, i.e.:
+    // enum Level {
+    //   None,        // -O0
+    //   Less,        // -O1
+    //   Default,     // -O2, -Os
+    //   Aggressive   // -O3
+    // };
+    machineBuilder->setCodeGenOptLevel((llvm::CodeGenOpt::Level)optLevel);
+    return machineBuilder->createTargetMachine();
+}
+
+/// Returns the cache level size from `targetInfo` for the `cacheLevel` provided. If `userCacheSize`
+/// is not zero, it returns `userCacheSize`.
+static unsigned getCacheLevelSize(llvm::TargetTransformInfo& targetInfo,
+                                  unsigned cacheLevel,
+                                  unsigned userCacheSize)
+{
+    if (userCacheSize)
+    {
+        return userCacheSize;
+    }
+
+    llvm::Optional<unsigned> optCacheLevelSize;
+    switch (cacheLevel)
+    {
+    case 1:
+        optCacheLevelSize = targetInfo.getCacheSize(llvm::TargetTransformInfo::CacheLevel::L1D);
+        break;
+    case 2:
+        optCacheLevelSize = targetInfo.getCacheSize(llvm::TargetTransformInfo::CacheLevel::L2D);
+        break;
+    default:
+        NGRAPH_UNREACHABLE("Unsupported cache level: ", cacheLevel, ". Only 1 and 2 are supported");
+    }
+
+    NGRAPH_CHECK(optCacheLevelSize.hasValue() && "Cache level size is not available in TTI");
+    return optCacheLevelSize.getValue();
+}
+
```

Comment:
See `runtime/cpu/cpu_backend.cpp` for what to do so things work with static linking as well as dynamically loaded libraries. It's mainly having the correctly-named extern "C" function available, which could just call your init here.

Reply:
I don't think this applies here since MLIRCPUBackend is not a replacement for `runtime/cpu/cpu_backend`. It is sub-component of the MLIR lib that gets linked into ngraph cpu backend. There is still more refactoring before we worry about static linking. Currently, Core, Backend and Runtime are all in one lib that we need to split up into 3. The Core will get linked to ngraph lib, and the rest into the ngraph backends.